Release pipeline

jweiser opened the pull request.

jweiser committed:
Add default 'gkb' directory

Historically, Reactome has had a '/usr/local/gkb/' and a '/usr/local/gkbdev' path.  The $gkb variable allows for both possibilities, though in practice they now resolve to the same directory.

jweiser committed:
removes unused Parallel::Loops module import

jweiser committed:
checks gk_central for instance

If the instance being checked exists in gk_central, it is known that it is NOT electronically inferred

jweiser committed:
stores the sorted array before returning

Addresses Perl::Critic complaint about unpredictable behaviour of returning the value of the sort function.
https://metacpan.org/pod/Perl::Critic::Policy::Subroutines::ProhibitReturnSort

jweiser committed:
Cleans use of white space

jweiser committed:
Adds semi-colon to end of catch block

jweiser committed:
Adds test for manually curated EWAS reused in electronic inferred RLE

jweiser committed:
Syntax bug fix

jweiser committed:
White space fixes

jweiser committed:
Adds backwards compatibility for regulation instances

Allows regulation instances prior to v65 still to be identified as electronically inferred

jweiser committed:
Bug fix - de-references array

jweiser committed:
Preserves order of hashes

jweiser committed:
Exports new functions from CommonUtils.pm

Exports in_gk_central($instance) and have_same_species($instance1, $instance2)

jweiser committed:
Moves Reactome2GO mapping file to archive

jweiser committed:
Refactoring and clean-up suggestions from Perl::Critic

jweiser committed:
Adds output file command line argument

jweiser committed:
Refactoring and style changes

jweiser committed:
Fixes whitespace

jweiser committed:
Refactoring and style changes

jweiser committed:
Cleans up style and Perl::Critic complaints

jweiser committed:
Ensures SQL dump files are zipped and moved to archive with other files

GZip files with .dump extension and move them as well as files with .err, .log, and .out extensions to the archive directory under the appropriate release version

jweiser committed:
Adds single "default sender" of pipeline e-mail

The 'from' field of automated release pipeline e-mails must be one and only one address.  The new "default sender" mail identifier is used for this.

jweiser committed:
Checks sender and recipients e-mails for validity

SolomonShorser-OICR commented on modules/GKB/Release/Step.pm:
```diff
@@ -197,474 +199,494 @@ use GKB::Release::Config;
+        }
+        $self->mail->{'body'} .= "$self->{name} step has completed successfully";
+    }
+    $self->mail_now();
```
Is there any way to catch errors if mailing fails?

jweiser replied:
Yup, I'm sure there's a way to do that.  I'll look into it.

jweiser replied:
Okay, I've added in some logic to capture mail errors and report them to the release log and on screen while running.  I've committed it and I just need to test it.

SolomonShorser-OICR replied:
:+1:

SolomonShorser-OICR commented on modules/GKB/Release/Utils.pm:
```diff
@@ -117,7 +117,7 @@ sub get_dba {
 sub set_environment {
     my $host = shift;
-    my $gkb = $hosts{$host};
+    my $gkb = $hosts{$host} // 'gkb';
```
I'm not sure why "gkb" is a good default for a host. Isn't the host supposed to refer to the hostname of the machine?

jweiser replied:
"gkb" is actually the directory alias.  The %hosts hash is a lookup from host to gkb alias as a holdover from when dev machines used 'gkbdev' and production machines used 'gkb'.  Release relevant machines should have the 'gkb' alias regardless now, but we may need to revise this function that sets environment to ensure it works with AWS and local machines.  

I'll put it on my to do list.

SolomonShorser-OICR replied:
Ok.

SolomonShorser-OICR commented on modules/GKB/CommonUtils.pm:
```diff
@@ -112,204 +114,259 @@ sub restore_database {
+}
+
+sub fetch_gk_central_physical_entity_lookup {
+    my $dba = get_dba('gk_central', 'reactomecurator.oicr.on.ca');
```
Should the hostname of this be configurable? If we are going to start running releases on local workstations, steps that rely on gk_central will probably perform better if we move a copy of gk_central to wherever the step's code is running.

jweiser replied:
Sure, I can have the function take an option for host and db

jweiser replied:
I changed the function to take db name and host as arguments, but I realized the function is actually called internally.  So you don't have to touch the source code, I have it use defaults imported from Config.pm for the curator database name and host.  Those values are taken from the Secrets.pm file.

SolomonShorser-OICR replied:
:+1:

SolomonShorser-OICR commented on modules/GKB/CommonUtils.pm:
```diff
@@ -112,204 +114,259 @@ sub restore_database {
+        return any { is_electronically_inferred($_)} @{$instance->reverse_attribute_value('catalystActivity')};
+    } elsif ($instance->is_a('Regulation')) {
+        # For backwards compatability when regulations contained the reaction like event (RLE) they regulated
+        # As of Reactome version 65, RLEs have contained their regulation instances in the 'regulatedBy' attribute
```
How long must this code support backwards compatibility? Will those regulations eventually be updated?

jweiser replied:
This is to allow queries to previous versions of release databases.  The regulations in those databases won't ever be updated I don't think because the data model they have doesn't support it.

SolomonShorser-OICR replied:
Ah, ok.

jweiser committed:
Makes curator database name and host configurable

jweiser committed:
Captures mail failures in release log

jweiser committed:
Removes unnecessary catch block

jweiser committed:
Updates capture of mail error messages

Ensures mail errors reported whether from exception or from populated package variable

SolomonShorser-OICR commented:
I don't have any other concerns with this code. There are merge conflicts in Step.pm, but as long as the merge doesn't cause any problems, then I'm OK with this.

jweiser committed:
Fixes arguments passed to reactome2go.pl

jweiser committed:
Bug fix -- corrects subroutine name

jweiser committed:
Bug fix -- corrects argument parsing

jweiser committed:
Specifies defaults output file name

uses version number or db name as suffix

cookersjs commented on modules/GKB/CommonUtils.pm:
```diff
@@ -112,204 +114,262 @@ sub restore_database {
+
+    # first, let's check to see if the "magic" note is in the Created InstanceEdit for this object.
+    my $created = $instance->created->[0];
+    if ($created && $created->note && $created->note eq "inferred events based on ensembl compara") {
```
Repeating the comment from the other PR: regex should check for the 'inferred events' string to account for when the statement includes Panther

jweiser replied:
Yup, I'll make a change.  I guess the note for Panther will be "inferred events based on panther"?

cookersjs commented on modules/GKB/CommonUtils.pm:
```diff
@@ -319,81 +379,76 @@ sub get_unique_species {
+
+    my @components;
+    if ($composite_entity) {
+        foreach my $component (
```
The format of this block takes a bit to understand. It might be easier if all of the arrays that will be iterated over were on the same line (instead of the two lines), or if all array elements were put into a single array before the for loop. This is obviously a style thing, I just felt it went against the code flow

jweiser replied:
Yup, that's a good point.  I'll change it.

cookersjs commented on modules/GKB/CommonUtils.pm:
```diff
@@ -319,81 +379,76 @@ sub get_unique_species {
+    my $author_instance;
+    try {
+        foreach my $modified_instance (reverse @{$event->modified}) {
+            $author_instance ||= $modified_instance->author->[0] unless $modified_instance->author->[0] && $modified_instance->author->[0]->db_id == 140537;
```
What is the magical number?

jweiser replied:
A magician never reveals his tricks...

I think it's Guanming's person id instance.  I'll rework it!

SolomonShorser-OICR replied:
Maybe a better question is: why are things authored by Guanming being treated differently?

jweiser replied:
Sorry, I should have explained.  

Guanming (and others like myself I may need to add in) are excluded because this function is meant to return the curator author and many modification instances are made by Guanming for data model purposes that Lisa requested be ignored.

I'll include a comment with the function and refactor to make the excluded authors more obvious.

cookersjs commented:
Done my review. Nothing jumped out so from my end you are good to merge

jweiser commented:
> Done my review. Nothing jumped out so from my end you are good to merge

Great, thanks.  I'll let you know when I make the changes before merging.

jweiser committed:
Bug fix -- variable name

jweiser committed:
Searches for inferred events note for ensembl compara or panther

jweiser committed:
Refactors multiple arrays in foreach loop to single function return

jweiser committed:
Refactors get_event_modifier

Comments added, simplifies logic, and makes exclusion of certain authors explicit

jweiser committed:
Removes export of low level function

jweiser committed:
Adds Esther to excluded curation event modifiers

jweiser committed:
Adds tests for event modifier name retrieval

cookersjs commented:
New changes look good, I believe you are good to merge

jweiser commented:
> New changes look good, I believe you are good to merge

Great, thanks!

jweiser committed:
Merge pull request #74 from reactome/release-pipeline/commit_goa

Release pipeline/commit goa

SolomonShorser-OICR reviewed (approved):
Looks good to me.

jweiser committed:
Merge branch 'master' into release_pipeline
